### PR TITLE
feat: rewrite upgrade-test.sh in rust

### PR DESCRIPTION
**Pull request by dhruv-1001**
Step towards completing #2101 (2c). This PR rewrites upgrade-test.sh in Rust.

### CODE REVIEW

File: fedimint-bin-tests/src/main.rs [hunk 1]
```diff
@@ -958,14 +958,97 @@ async fn reconnect_test(dev_fed: DevFed, process_mgr: &ProcessManager) -> Result
...
+    fed.start_server(process_mgr, 2, Some(upgrade_epoch))
+        .await?;
+    fed.start_server(process_mgr, 3, Some(upgrade_epoch))
+        .await?;
+
+    info!("Federation started");
+
+    bitcoind.mine_blocks(1).await?;
+    fed.await_block_sync().await?;
+    fed.await_all_peers().await?;
```
Comment (dhruv-1001) [inline fedimint-bin-tests/src/main.rs:1047]: Having some problem with `await_all_peers`
`command: /home/dhruv/development/oss/forked/fedimint/target/debug/fedimint-cli --data-dir=/tmp/nix-shell.UEIDMP/fm-EhYd/cfg api /module/2/block_height
stdout:

stderr:
2023-04-19T14:06:58.194586Z ERROR net::api: unable to connect to server err=Networking or low-level protocol error: Error when opening the TCP socket: Connection refused (os error 111)`

Comment (dhruv-1001) [reply fedimint-bin-tests/src/main.rs:1047]: The Federation seems to be running, i'm not sure why this command is failing. Any possible reason for this failure?

**Comment by jkitman** [comment]
@dhruv-1001 sorry you spent time on this I should have caught it earlier, but it's going to be tested as part of the new config gen API probably.

**Comment by dhruv-1001** [comment]
> @dhruv-1001 sorry you spent time on this I should have caught it earlier, but it's going to be tested as part of the new config gen API probably.
> 
> Still we could merge your code until that is finished.

Oh okay, I'm almost done with this PR.

File: fedimint-bin-tests/src/main.rs (hunk 1 above)
Comment (dhruv-1001) [reply fedimint-bin-tests/src/main.rs:1040]: I was killing `fedimintd`, instead I should have waited the it's shutdown.

File: fedimint-bin-tests/src/main.rs [hunk 2]
```diff
@@ -966,6 +966,82 @@ async fn reconnect_test(dev_fed: DevFed, process_mgr: &ProcessManager) -> Result
...
+    )
+    .env("FM_PASSWORD", "pass2")
+    .run()
+    .await?;
+
+    info!("### Wait for server shutdowns");
+    fed.await_server_shutdown(0).await?;
+    fed.await_server_shutdown(1).await?;
+    fed.await_server_shutdown(2).await?;
+    fed.await_server_shutdown(3).await?;
```
Comment (Maan2003) [inline fedimint-bin-tests/src/main.rs:1031]: do all these in parallel?

**Comment by dhruv-1001** [comment]
Hey @Maan2003, can you re-run the `Code coverage` check? Thanks!

**Comment by Maan2003** [comment]
needs rebase

File: fedimint-bin-tests/src/util.rs [hunk 3]
```diff
@@ -24,17 +24,28 @@ pub struct ProcessHandle(Arc<ProcessHandleInner>);
...
+        let mut child = std::mem::take(&mut process_handle_inner.child).unwrap();
+        child.wait().await?;
+        Ok(())
+    }
+
+    pub fn get_inner(self) -> Result<ProcessHandleInner> {
+        let arc_process_handle_inner = self.0;
+        match Arc::try_unwrap(arc_process_handle_inner) {
+            Ok(process_handler_inner) => Ok(process_handler_inner),
+            Err(_) => return Err(anyhow!("Cannot kill process because of clones")),
```
Comment (douglaz) [inline fedimint-bin-tests/src/util.rs:46]:
```suggestion
            Err(_) => anyhow::bail!("Cannot kill process because of clones"),
```

**Comment by dhruv-1001** [comment]
No longer needed.

**Pull request closed without merging**